### ISSUE TITLE
Jacobians of step function

### ISSUE THREAD

**Issue by proceduralia** [google/brax#83]
Hi,

In the example about analytic policy gradients, the gradient w.r.t. the policy parameters is computed without explicit access to the Jacobians of the dynamics.

If another model-based method requires explicit use of the Jacobians of the dynamics, what would be the recommended way to handle them? Calling the jax Jacobian functions on the environment's step function can give the one w.r.t. the simulator's state, not the observations used by a policy. What is the best way to compute those Jacobians, i.e., the derivative of the dynamics w.r.t. observation and action?

Thanks,

Pierluca

**Comment by cdfreeman-google** [comment]
This is a little bit tricky because the observations aren't necessarily invertible from the state.  That is, we typically have:

```
f(state, action) -> state
```

where observations are collected inside the step function after an action has occurred.  If we had an invertible h(s) = o, then we could of course just replace state with `o` and immediately take h^{-1}(o) = s inside the step function, but, again, this isn't really guaranteed.

I suspect what you _really_ want, though, is the derivative of the step function with respect to the observation of a policy, or, something like this:

```
def step_obs(brax_env, policy, state, obs):
  action = policy(obs)
  qp, info = brax_env.sys.step(state.qp, action)
  obs = brax_env._get_obs(qp, info)
  return obs
```

Where this guy's jacobian is:
```
jac_step_obs = jax.jit(jax.jacobian(step_obs, argnums=3), static_argnums=0)
```

This is of course specific to the dynamics generated by the particular `policy` you feed in, but that might be enough?

**Comment by cdfreeman-google** [comment]
Closing this--feel free to reopen if this did not answer your question!
